src/overlays/actors/ovl_En_Racedog/z_en_racedog.c

```diff
@@ -399,66 +508,90 @@ void func_80B24F08(EnRacedog* this) {
+ * sFurthestPoint and sFirstPlaceIndex variables.
+ */
+void EnRacedog_UpdateRaceVariables(EnRacedog* this) {
+    if ((this->currentPoint >= 9) && (this->raceStatus == RACEDOG_RACE_STATUS_BEFORE_POINT_9)) {
+        this->raceStatus = RACEDOG_RACE_STATUS_BETWEEN_POINT_9_AND_11;
     }
 
-    if ((this->unk_1E8 >= 0xB) && (this->unk_29C == 1)) {
-        this->unk_29C = 2;
+    if ((this->currentPoint >= 11) && (this->raceStatus == RACEDOG_RACE_STATUS_BETWEEN_POINT_9_AND_11)) {
+        this->raceStatus = RACEDOG_RACE_STATUS_AFTER_POINT_11;
     }
 
-    if (((this->unk_1E8 >= D_80B25D44) || (this->unk_29C <= 0)) && (this->unk_1E8 > D_80B25D44)) {
-        D_80B25D44 = this->unk_1E8;
-        D_80B25D4C = this->unk_290;
+    if (((this->currentPoint >= sFurthestPoint) || (this->raceStatus <= RACEDOG_RACE_STATUS_BEFORE_POINT_9)) &&
+        (this->currentPoint > sFurthestPoint)) {
+        sFurthestPoint = this->currentPoint;
+        sFirstPlaceIndex = this->index;
     }
 }
 
-void func_80B2538C(EnRacedog* this) {
-    if (func_80B25490(this, D_80B25E78) && this->unk_29C == 2) {
-        D_80B25D40++;
-        if (D_80B25D40 == 1) {
+/**
+ * Checks to see if this dog has finished the race. This function is responsible for changing the
+ * music when the first dog finishes the race, and it is also responsible for updating the event
+ * flags with what position the player's selected dog finished in.
+ */
+void EnRacedog_CheckForFinish(EnRacedog* this) {
+    if (EnRacedog_IsOverFinishLine(this, sFinishLineCoordinates) &&
+        this->raceStatus == RACEDOG_RACE_STATUS_AFTER_POINT_11) {
+        sNumberOfDogsFinished++;
+        if (sNumberOfDogsFinished == 1) {
             Audio_QueueSeqCmd(NA_BGM_HORSE_GOAL | 0x8000);
             play_sound(NA_SE_SY_START_SHOT);
         }
 
-        this->unk_29C = 3;
-        if (this->unk_290 == this->unk_292) {
-            gSaveContext.eventInf[0] = (gSaveContext.eventInf[0] & 7) | (D_80B25D40 * 8);
+        this->raceStatus = RACEDOG_RACE_STATUS_FINISHED;
+        if (this->index == this->selectedDogIndex) {
+            gSaveContext.eventInf[0] = (gSaveContext.eventInf[0] & 7) | (sNumberOfDogsFinished * 8);
         }
     }
 }
 
-void func_80B25448(EnRacedog* this) {
+/**
+ * Slows the dog's running speed animation if its speed is less than 3.0, and sets it
+ * to normal speed otherwise.
+ */
+void EnRacedog_UpdateRunAnimationPlaySpeed(EnRacedog* this) {
     if (this->actor.speedXZ < 3.0f) {
-        D_80B25EF0[2].playSpeed = 0.9f;
+        sAnimations[RACEDOG_ANIMATION_RUN].playSpeed = 0.9f;
     } else {
-        D_80B25EF0[2].playSpeed = 1.0f;
+        sAnimations[RACEDOG_ANIMATION_RUN].playSpeed = 1.0f;
     }
 }
 
-s32 func_80B25490(EnRacedog* this, Vec2f* arg1) {
+/**
+ * Returns true if the dog's current position is in the finish line.
+ */
+s32 EnRacedog_IsOverFinishLine(EnRacedog* this, Vec2f* finishLineCoordinates) {
```

Comment:
Yeah that's basically what's going on here. It's doing multiple cross/wedge products to determine whether the dog is inside the finish line. It might be easier to show with diagrams.

This diagram shows what the first `if` is checking; it returns false if the dog is in the red sections. The yellow box is a box drown around the finish line coordinates, and the arrow is the direction the dogs approach the finish line:
![image](https://user-images.githubusercontent.com/12245827/162556574-a6a52653-5491-4c56-bfa8-a4410b977c68.png)

This diagram shows what the second `if` is checking:
![image](https://user-images.githubusercontent.com/12245827/162556597-474240e5-a777-44ea-90a9-b00669d5afbb.png)

This diagram shows what the third `if` is checking:
![image](https://user-images.githubusercontent.com/12245827/162556610-2f6cb579-ea94-4b39-a031-f091f5af84fe.png)

There's only a single segment that is blue in all three of these, and it's the area within the finish line coordinates. I'm not sure how to make the *documentation* better, but hopefully that explains what's going on here and what all this math represents.

Reply:
I think for me the main thing that needs clarifying is that the rectangle it's checking is a horizontal one that roughly lines up with the finish line texture.

Reply:
That's correct. The entire racetrack is built slanted, so it's *really* a parallelogram. But from the perspective of the track itself, it's just a simple rectangle that lines up roughly with the texture on the ground.